test_programs/compile_success_empty/associated_types/src/main.nr

```diff
@@ -0,0 +1,28 @@
+trait Collection {
+    type Elem;
+
+    fn cget(self, index: u32) -> Option<Self::Elem>;
```

Comment:
Two things I noticed:
- If you change `Self::Elem` to `Self::Elem2` you get an error saying "Could not resolve Self in path"
- If you change `Self::Elm` to `Elem`, it still compiles fine (but maybe it shouldn't, like in Rust?) That said, changing the one below does give an error

Reply:
Yes, I think these are existing issues with how we handle `Self` in general. The second probably shouldn't compile... but seems okay to allow I think. It's a result of us treating traits as a pseudo-module so Elem is in scope to cget.